File: Frontend/blockchain_test.java
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import com.google.gson.Gson;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import spark.Spark;

class Block {
    private int index;
    private String timestamp;
    private int proof;
    private String previousHash;

    public Block(int index, String timestamp, int proof, String previousHash) {
        this.index = index;
        this.timestamp = timestamp;
        this.proof = proof;
        this.previousHash = previousHash;
    }

    // Getters
    public int getIndex() { return index; }
    public String getTimestamp() { return timestamp; }
    public int getProof() { return proof; }
    public String getPreviousHash() { return previousHash; }
}

class Blockchain {
    private List<Block> chain;
    private Gson gson = new Gson();

    public Blockchain() {
        this.chain = new ArrayList<>();
        createBlock(1, "0");
    }

    public Block createBlock(int proof, String previousHash) {
        Block block = new Block(
            chain.size() + 1,
            new Date().toString(),
            proof,
            previousHash
        );
        chain.add(block);
        return block;
    }

    public Block getPreviousBlock() {
        return chain.get(chain.size() - 1);
    }

    public int proofOfWork(int previousProof) {
        int newProof = 1;
        boolean checkProof = false;

        while (!checkProof) {
            String hashOperation = sha256(String.valueOf(newProof * newProof - previousProof * previousProof));
            if (hashOperation.substring(0, 5).equals("00000")) {
                checkProof = true;
            } else {
                newProof++;
            }
        }

        return newProof;
    }

    public String hash(Block block) {
        String json = gson.toJson(block);
        return sha256(json);
    }

    public boolean isChainValid(List<Block> chain) {
        Block previousBlock = chain.get(0);
        int blockIndex = 1;

        while (blockIndex < chain.size()) {
            Block block = chain.get(blockIndex);
            if (!block.getPreviousHash().equals(hash(previousBlock))) {
                return false;
            }

            int previousProof = previousBlock.getProof();
            int proof = block.getProof();
            String hashOperation = sha256(String.valueOf(proof * proof - previousProof * previousProof));

            if (!hashOperation.substring(0, 5).equals("00000")) {
                return false;
            }
            previousBlock = block;
            blockIndex++;
        }

        return true;
    }

    private String sha256(String input) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes());
            StringBuilder hexString = new StringBuilder();

            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }

            return hexString.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Blockchain blockchain = new Blockchain();
        
        Spark.port(5000);
        
        Spark.get("/mine_block", (req, res) -> {
            Block previousBlock = blockchain.getPreviousBlock();
            int previousProof = previousBlock.getProof();
            int proof = blockchain.proofOfWork(previousProof);
            String previousHash = blockchain.hash(previousBlock);
            Block block = blockchain.createBlock(proof, previousHash);
            
            String response = String.format("{\"message\":\"A block is MINED\",\"index\":%d,\"timestamp\":\"%s\",\"proof\":%d,\"previous_hash\":\"%s\"}",
                block.getIndex(), block.getTimestamp(), block.getProof(), block.getPreviousHash());
            
            res.type("application/json");
            return response;
        });
        
        Spark.get("/get_chain", (req, res) -> {
            List<Block> chain = blockchain.getChain();
            String response = String.format("{\"chain\":%s,\"length\":%d}",
                new Gson().toJson(chain), chain.size());
            
            res.type("application/json");
            return response;
        });
        
        Spark.get("/valid", (req, res) -> {
            boolean valid = blockchain.isChainValid(blockchain.getChain());
            String response;
            
            if (valid) {
                response = "{\"message\":\"The Blockchain is valid.\"}";
            } else {
                response = "{\"message\":\"The Blockchain is not valid.\"}";
            }
            
            res.type("application/json");
            return response;
        });
    }
}

File: Frontend/blockchain_test.js
const express = require('express');
const crypto = require('crypto');
const bodyParser = require('body-parser');

class Block {
    constructor(index, timestamp, proof, previousHash) {
        this.index = index;
        this.timestamp = timestamp;
        this.proof = proof;
        this.previousHash = previousHash;
    }
}

class Blockchain {
    constructor() {
        this.chain = [];
        this.createBlock(1, '0');
    }

    createBlock(proof, previousHash) {
        const block = new Block(
            this.chain.length + 1,
            new Date().toString(),
            proof,
            previousHash
        );
        this.chain.push(block);
        return block;
    }

    getPreviousBlock() {
        return this.chain[this.chain.length - 1];
    }

    proofOfWork(previousProof) {
        let newProof = 1;
        let checkProof = false;

        while (!checkProof) {
            const hashOperation = crypto
                .createHash('sha256')
                .update((newProof ** 2 - previousProof ** 2).toString())
                .digest('hex');
            
            if (hashOperation.substring(0, 5) === '00000') {
                checkProof = true;
            } else {
                newProof++;
            }
        }

        return newProof;
    }

    hash(block) {
        const blockString = JSON.stringify(block);
        return crypto
            .createHash('sha256')
            .update(blockString)
            .digest('hex');
    }

    isChainValid(chain) {
        let previousBlock = chain[0];
        let blockIndex = 1;

        while (blockIndex < chain.length) {
            const block = chain[blockIndex];
            if (block.previousHash !== this.hash(previousBlock)) {
                return false;
            }

            const previousProof = previousBlock.proof;
            const proof = block.proof;
            const hashOperation = crypto
                .createHash('sha256')
                .update((proof ** 2 - previousProof ** 2).toString())
                .digest('hex');

            if (hashOperation.substring(0, 5) !== '00000') {
                return false;
            }

            previousBlock = block;
            blockIndex++;
        }

        return true;
    }
}

const app = express();
app.use(bodyParser.json());

const blockchain = new Blockchain();

// Mine a new block
app.get('/mine_block', (req, res) => {
    const previousBlock = blockchain.getPreviousBlock();
    const previousProof = previousBlock.proof;
    const proof = blockchain.proofOfWork(previousProof);
    const previousHash = blockchain.hash(previousBlock);
    const block = blockchain.createBlock(proof, previousHash);

    const response = {
        message: 'A block is MINED',
        index: block.index,
        timestamp: block.timestamp,
        proof: block.proof,
        previous_hash: block.previousHash
    };

    res.status(200).json(response);
});

// Get the full blockchain
app.get('/get_chain', (req, res) => {
    const response = {
        chain: blockchain.chain,
        length: blockchain.chain.length
    };
    res.status(200).json(response);
});

// Check if the blockchain is valid
app.get('/valid', (req, res) => {
    const valid = blockchain.isChainValid(blockchain.chain);
    if (valid) {
        res.status(200).json({ message: 'The Blockchain is valid.' });
    } else {
        res.status(200).json({ message: 'The Blockchain is not valid.' });
    }
});

// Run the server
const PORT = 5000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});